src/Symfony/Component/Config/Definition/ArrayNode.php

```diff
@@ -52,17 +53,41 @@ public function setNormalizeKeys($normalizeKeys)
      */
     protected function preNormalize($value)
     {
-        if (!$this->normalizeKeys || !is_array($value)) {
+        if (!is_array($value)) {
             return $value;
         }
 
-        $normalized = array();
+        $normalized = $value;
+        if ($this->normalizeKeys) {
+            $normalized = array();
+
+            foreach ($value as $k => $v) {
+                if (false !== strpos($k, '-') && false === strpos($k, '_') && !array_key_exists($normalizedKey = str_replace('-', '_', $k), $value)) {
+                    $normalized[$normalizedKey] = $v;
+                } else {
+                    $normalized[$k] = $v;
+                }
+            }
+        }
+
+        foreach ($this->deprecations as $deprecation) {
+            list($names, $handler, $message) = $deprecation;
+
+            $given = array_intersect(array_keys($normalized), $names);
+            if (!$given) {
+                continue;
+            }
 
-        foreach ($value as $k => $v) {
-            if (false !== strpos($k, '-') && false === strpos($k, '_') && !array_key_exists($normalizedKey = str_replace('-', '_', $k), $value)) {
-                $normalized[$normalizedKey] = $v;
-            } else {
-                $normalized[$k] = $v;
+            if (null === $message) {
+                $message = sprintf(1 === count($given) ? 'The node "%s" is deprecated at path "%s".' : 'The nodes "%s" are deprecated at path "%s".',  implode('", "', $given), $this->getPath());
+            }
+            @trigger_error($message, E_USER_DEPRECATED);
+
+            if (null !== $handler) {
+                $normalized = $handler($this->remapXml($normalized));
```

Comment:
Oh, and on a side note; it was a really weird experience noticing that xml remapping is applied after any normalization.

I totally expected this to be done _before_ normalization. Like normalizing keys...